osquery/worker/ipc/posix/pipe_channel_factory.cpp

```diff
@@ -0,0 +1,81 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed in accordance with the terms specified in
+ *  the LICENSE file found in the root directory of this source tree.
+ */
+
+#include "pipe_channel_factory.h"
+
+#include <unistd.h>
+
+#include <array>
+#include <stdexcept>
+#include <string>
+
+namespace osquery {
+
+PipeChannelTicket::PipeChannelTicket(std::array<int, 2> read_pipe_fds,
+                                     std::array<int, 2> write_pipe_fds)
+    : read_pipe_fds_(read_pipe_fds), write_pipe_fds_(write_pipe_fds) {}
+
+PipeChannelTicket PipeChannelFactory::createChannelTicket() {
+  std::array<int, 2> read_pipe_fds{};
+  auto result = pipe(read_pipe_fds.data());
+
+  if (result == -1) {
+    throw std::runtime_error(
+        "Failed to create parent_write_child_read_pipe, error: " +
+        std::to_string(errno));
+  }
+
+  std::array<int, 2> write_pipe_fds{};
+  result = pipe(write_pipe_fds.data());
```

Comment:
If the first call to `pipe` succeeds and this one fails, will we leak a pipe?

Reply:
Good catch! 
I've also found a second bug thanks to this in PipeChannelTicket, that when being destroyed without pipes internally set, it would close the fd 0; fixed that too.